# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLMemoryTracker.h
# Commit: ccfa767dba64
# Full Hash: ccfa767dba644dc193e0246eb8e8ff3377e8b8a5
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Client-side bindings mirror for precise CC, and merge similar codepaths. r=handyman
#   
#   * Context loss using RAII
#   * Move Program reflection Client-side
#   
# ==============================================================================

diff -r 71c122ac0ca7 -r ccfa767dba64 dom/canvas/WebGLMemoryTracker.h
--- a/dom/canvas/WebGLMemoryTracker.h	Wed Jan 08 22:19:14 2020 +0000
+++ b/dom/canvas/WebGLMemoryTracker.h	Wed Jan 08 22:19:16 2020 +0000
@@ -8,46 +8,45 @@
 
 #include "mozilla/StaticPtr.h"
 #include "nsIMemoryReporter.h"
+#include <unordered_set>
 
 namespace mozilla {
 
-class WebGLContext;
+class HostWebGLContext;
 
 class WebGLMemoryTracker : public nsIMemoryReporter {
   NS_DECL_THREADSAFE_ISUPPORTS
   NS_DECL_NSIMEMORYREPORTER
 
-  WebGLMemoryTracker();
   static StaticRefPtr<WebGLMemoryTracker> sUniqueInstance;
 
+  static RefPtr<WebGLMemoryTracker> Create();
+
+  static RefPtr<WebGLMemoryTracker> Get() {
+    if (!sUniqueInstance) {
+      sUniqueInstance = WebGLMemoryTracker::Create().get();
+    }
+    return sUniqueInstance;
+  }
+
   // Here we store plain pointers, not RefPtrs: we don't want the
   // WebGLMemoryTracker unique instance to keep alive all
   // WebGLContexts ever created.
-  typedef nsTArray<const WebGLContext*> ContextsArrayType;
-  ContextsArrayType mContexts;
-
-  void InitMemoryReporter();
-
-  static WebGLMemoryTracker* UniqueInstance();
-
-  static ContextsArrayType& Contexts() { return UniqueInstance()->mContexts; }
-
-  friend class WebGLContext;
+  std::unordered_set<const HostWebGLContext*> mContexts;
 
  public:
-  static void AddWebGLContext(const WebGLContext* c) {
-    Contexts().AppendElement(c);
+  static void AddContext(const HostWebGLContext* c) {
+    const auto tracker = Get();
+    tracker->mContexts.insert(c);
   }
 
-  static void RemoveWebGLContext(const WebGLContext* c) {
-    ContextsArrayType& contexts = Contexts();
-    contexts.RemoveElement(c);
-    if (contexts.IsEmpty()) {
-      sUniqueInstance = nullptr;
-    }
+  static void RemoveContext(const HostWebGLContext* c) {
+    const auto tracker = Get();
+    tracker->mContexts.erase(c);
   }
 
  private:
+  WebGLMemoryTracker();
   virtual ~WebGLMemoryTracker();
 
   static int64_t GetTextureMemoryUsed();
@@ -68,7 +67,7 @@
 
   static int64_t GetShaderCount();
 
-  static int64_t GetContextCount() { return Contexts().Length(); }
+  static int64_t GetContextCount() { return Get()->mContexts.size(); }
 };
 
 }  // namespace mozilla